cpp/src_prims/matrix/batched_matrix.hpp

```diff
@@ -0,0 +1,503 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <functional>
+#include <stdexcept>
+#include <tuple>
+#include <unordered_map>
+#include <vector>
+
+#include <linalg/binary_op.h>
+#include <linalg/cublas_wrappers.h>
+#include <memory>
+
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+
+#include <common/cumlHandle.hpp>
+#include <cuml/common/utils.hpp>
+#include <cuml/cuml.hpp>
+
+namespace MLCommon {
+namespace Matrix {
+
+/**
+ * @brief An allocation function for `BatchedMatrixMemory`.
+ * 
+ * @note Written as a free function because I had trouble getting the
+ *       __device__ lambda to compile as a member function of the
+ *       `BatchedMatrixMemory` struct.
+ * 
+ * @param[in]  shape        Shape of each matrix (rows, columns)
+ * @param[in]  num_batches  Number of matrices in the batch
+ * @param[in]  setZero      Whether to initialize the allocated matrix with all zeros
+ * @param[in]  allocator    Device memory allocator
+ * @param[in]  stream       CUDA stream
+ * 
+ * @return Pair (A_dense, A_array): pointer to the raw data and pointer
+ *         to the array of pointers to each matrix in the batch
+ */
+std::pair<double*, double**> BMM_Allocate(
```

Comment:
The functions and class in this header file are not linking-friendly: if two source files include this header they can't be linked together because of *multiple definition* conflicts. A good fix that will be useful later anyway is to make the whole class templated.

I can PR this on your branch if you want, I need it for the arima parameters initialization work anyway.

Reply:
That would be really great. I guess it won't take too long.

Reply:
Just finished the template and adjusted the tests to also test single precision.